Explicit Resource Management article added

bagherani opened the pull request:
This PR will add Mohi Bagherani as a new author and `/articles/explicit-resource-management-in-typescript-5.2` article.

bagherani committed:
Explicit Resource Management article added

DaveBitter commented on data/blog/explicit-resource-management-in-typescript-5.2.md:
```diff
@@ -0,0 +1,229 @@
+
+# ECMAScript Explicit Resource Management early implementation in Typescript 5.2
+
+Explicit Resource Management indicates a system whereby the lifetime of a “resource” is managed explicitly by the user either imperatively (by directly calling a method like Symbol.dispose) or declaratively (through a block-scoped declaration like using).
```
I think this reads a bit easier:

Explicit Resource Management indicates a system whereby the lifetime of a “resource” is managed explicitly by the user. This can be done either imperatively (by directly calling a method like Symbol.dispose) or declaratively (through a block-scoped declaration like using).

bagherani replied:
Done

DaveBitter commented on data/blog/explicit-resource-management-in-typescript-5.2.md:
```diff
@@ -0,0 +1,229 @@
+
+Explicit Resource Management indicates a system whereby the lifetime of a “resource” is managed explicitly by the user either imperatively (by directly calling a method like Symbol.dispose) or declaratively (through a block-scoped declaration like using).
+
+A **Resource** is an object with a specific lifetime, at the end of which either a lifetime-sensitive operation should be performed or a non-garbage-collected reference (such as a file handle, socket, etc.) should be closed or freed.
```
This sentence is quite difficult to read. Can you try splitting it up so it reads a bit easier?

bagherani replied:
Simplified.

DaveBitter commented on data/blog/explicit-resource-management-in-typescript-5.2.md:
```diff
@@ -0,0 +1,229 @@
+
+At the moment of writing this article, the ECMAScript [proposal](https://github.com/tc39/proposal-explicit-resource-management) for Explicit Resource Management is at [stage 3 ](https://tc39.es/process-document/)which means it will be launched within a few months and this is the time that the TypeScript community has already implemented it in Typescript [version 5.2](https://devblogs.microsoft.co ...
+
+This implementation will add a feature to the javascript language that we had similar to in other languages such as C# [using](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/statements#1314-the-using-statement) syntax, Python [with](https://docs.python.org/3/reference/compound_stmts.html#the-with-statement) syntax, or [try](https://docs.oracle.com/javase ...
```
javascript -> JavaScript

bagherani replied:
Fixed

DaveBitter commented on data/blog/explicit-resource-management-in-typescript-5.2.md:
````diff
@@ -0,0 +1,229 @@
+// ...
+```
+
+Later when the execution of the program goes in outside scope of the place the obj has been defined, the `dispose` method will be called.
````
This sentence is a bit difficult to read

bagherani replied:
Simplified

DaveBitter commented on data/blog/explicit-resource-management-in-typescript-5.2.md:
```diff
@@ -0,0 +1,229 @@
+
+Later when the execution of the program goes in outside scope of the place the obj has been defined, the `dispose` method will be called.
+
+To tell the Typescript compiler how to translate the new `using` syntax into a `try/finally` that the current Javascript can interpret, Typescript came up with an interface called Disposable that can be implemented by your class that makes you to have a `[Symbol.dispose]` method in that class like below:
```
Javascript -> JavaScript
Typescript -> TypeScript

It's probably best to go over the entire article and make sure that you are consistent

bagherani replied:
All done.

DaveBitter commented on data/blog/explicit-resource-management-in-typescript-5.2.md:
```diff
@@ -0,0 +1,229 @@
+
+Later when the execution of the program goes in outside scope of the place the obj has been defined, the `dispose` method will be called.
+
+To tell the Typescript compiler how to translate the new `using` syntax into a `try/finally` that the current Javascript can interpret, Typescript came up with an interface called Disposable that can be implemented by your class that makes you to have a `[Symbol.dispose]` method in that class like below:
```
This sentence is quite long and difficult to read, can you split this up in multiple sentences?

bagherani replied:
Thanks. done.

DaveBitter commented on data/blog/explicit-resource-management-in-typescript-5.2.md:
````diff
@@ -0,0 +1,229 @@
+// ...
+```
+
+Or if you have a **function** instead of a **class**, your function should return and object that has the `Symbol.dispose` method:
````
Or, if you have...

Grammarly usually highlights commas like this. Can you run the article through Grammarly?

bagherani replied:
Fixed, also I fixed some grammar issues with Grammarly too.

DaveBitter commented on data/blog/explicit-resource-management-in-typescript-5.2.md:
```diff
@@ -0,0 +1,229 @@
+
+### DisposableStack class
+
+Implementing the `Symbol.dispose` can be considered as a good pattern if you have a class or a function that needs to be disposed at someplace, but here is two problems: First, in some cases implementing this method might add too much abstraction to your code and second, there are a lot of libraries and modules that haven’t implemented the dispose method it yet. So, for using them, you have to wr ...
```
This sentence is quite long. Can you split this up?

bagherani replied:
Done.

DaveBitter commented on data/blog/explicit-resource-management-in-typescript-5.2.md:
````diff
@@ -0,0 +1,229 @@
+}
+```
+
+This class has been implemented (with all of its sophistications that might have later) to only have a dispose method that would be called by the engine, but there is a simpler solution that Typescript comes up with: the `DisposableStack` class.
````
that **it** might have later

bagherani replied:
Fixed.

DaveBitter commented on data/blog/explicit-resource-management-in-typescript-5.2.md:
````diff
@@ -0,0 +1,229 @@
+}
+```
+
+Here we’ve created an instance of `DisposableStack` right after opening the file; It has the defer method that accepts a callback function which is suitable for clean-up purpose that will be invoked at the moment of disposing the cleanup object. The compiled version of the above code will be in order that at the finally block the callback function passed to the defer method will be called.
````
clean-up purpose**s**

bagherani replied:
Fixed

DaveBitter commented on data/blog/explicit-resource-management-in-typescript-5.2.md:
````diff
@@ -0,0 +1,229 @@
+}
+```
+
+Here we’ve created an instance of `DisposableStack` right after opening the file; It has the defer method that accepts a callback function which is suitable for clean-up purpose that will be invoked at the moment of disposing the cleanup object. The compiled version of the above code will be in order that at the finally block the callback function passed to the defer method will be called.
````
You use cleanup and clean-up

bagherani replied:
All fixed.

DaveBitter commented on data/blog/explicit-resource-management-in-typescript-5.2.md:
````diff
@@ -0,0 +1,229 @@
+}
+```
+
+Here we’ve created an instance of `DisposableStack` right after opening the file; It has the defer method that accepts a callback function which is suitable for clean-up purpose that will be invoked at the moment of disposing the cleanup object. The compiled version of the above code will be in order that at the finally block the callback function passed to the defer method will be called.
````
The last sentence is a bit difficult to read and I think isn't grammatically correct

bagherani replied:
Rewritten.

DaveBitter commented on data/blog/explicit-resource-management-in-typescript-5.2.md:
```diff
@@ -0,0 +1,229 @@
+theme: 'blue'
+canonicalUrl: 'https://medium.com/@bagherani/ecmascript-explicit-resource-management-early-implementation-in-typescript-5-2-5e4d08b2aee3'
+---
+
```
Can you add a section on why this is a useful addition to TypeScript and why the reader should learn about it?

bagherani replied:
A motivation section has been added to the top of the article.

bagherani committed:
Grammars fixed and sentences simplified

remypar5 commented on data/blog/explicit-resource-management-in-typescript-5.2.md:
```diff
@@ -0,0 +1,228 @@
+
+This article on ECMAScript Explicit Resource Management implementation in TypeScript 5.2 should be learned by developers who are interested in managing resources explicitly in their JavaScript programs. It explains how this feature works and how it can simplify resource management code.
+
+Explicit Resource Management indicates a system whereby the lifetime of a “resource” is managed explicitly by the user. This can be done either imperatively (by directly calling a method like `Symbol.dispose`) or declaratively (through a block-scoped declaration like using).
```
```suggestion
Explicit Resource Management indicates a system whereby the lifetime of a “resource” is managed explicitly by the user. This can be done either imperatively (by directly calling a method like `Symbol.dispose`) or declaratively (through a block-scoped declaration like `using`).

Highlight code keywords as such
```

remypar5 commented on data/blog/explicit-resource-management-in-typescript-5.2.md:
```diff
@@ -0,0 +1,228 @@
+
+At the moment of writing this article, the ECMAScript [proposal](https://github.com/tc39/proposal-explicit-resource-management) for Explicit Resource Management is at [stage 3](https://tc39.es/process-document/) which means it will be launched within a few months and this is the time that the TypeScript community has already implemented it in TypeScript [version 5.2](https://devblogs.microsoft.co ...
+
+This implementation will add a feature to the JavvaScript language that we had similar to in other languages such as C# [using](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/statements#1314-the-using-statement) syntax, Python [with](https://docs.python.org/3/reference/compound_stmts.html#the-with-statement) syntax, or [try](https://docs.oracle.com/javas ...
```
```suggestion
This implementation will add a feature to the JavvaScript language that we had similar to in other languages such as C# [`using`](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/statements#1314-the-using-statement) syntax, Python [`with`](https://docs.python.org/3/reference/compound_stmts.html#the-with-statement) syntax, or [`try`](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html) in Java.
```

remypar5 commented on data/blog/explicit-resource-management-in-typescript-5.2.md:
```diff
@@ -0,0 +1,228 @@
+
+When the program's execution goes outside of the block that the `obj` has been defined, the object will no longer be needed and the `dispose` method will be called.
+
+TypeScript introduced an interface called `Disposable` that can be used to implement the using syntax. To do this, your class should implement the `Disposable` interface, which requires the class to have a [Symbol.dispose] method. This method is responsible for freeing up any resources used by the class when it is no longer needed. By implementing the `Disposable` interface, the TypeScript compil ...
```
```suggestion
TypeScript introduced an interface called `Disposable` that can be used to implement the using syntax. To do this, your class should implement the `Disposable` interface, which requires the class to have a `[Symbol.dispose]` method. This method is responsible for freeing up any resources used by the class when it is no longer needed. By implementing the `Disposable` interface, the TypeScript compiler knows how to translate the using syntax into a `try/finally` block that the current JavaScript engine can understand.
```

remypar5 commented on data/blog/explicit-resource-management-in-typescript-5.2.md:
````diff
@@ -0,0 +1,228 @@
+}
+```
+
+This class has been implemented (with all of its sophistication that it might have later) to only have a `dispose` method that would be called by the engine, but there is a simpler solution that TypeScript comes up with the `DisposableStack` class.
````
```suggestion
This class has been implemented (with all of its sophistication that it might have later) to only have a `Symbol.dispose` method that would be called by the engine, but there is a simpler solution that TypeScript comes up with the `DisposableStack` class.
```

remypar5 replied:
```suggestion
This class has been implemented (with all of its sophistication that it might have later) to only have a `dispose` method that would be called by the engine. However, TypeScript 5.2 introduces the `DisposableStack` class, which makes it simpler to use Explicit Resource Management.
```

remypar5 commented on data/blog/explicit-resource-management-in-typescript-5.2.md:
````diff
@@ -0,0 +1,228 @@
+}
+```
+
+Here we’ve created an instance of `DisposableStack` right after opening the file; It has the defer method that accepts a callback function which is suitable for clean-up that will be invoked at the moment of disposing the clean-up object.  
````
```suggestion
Here we’ve created an instance of `DisposableStack` right after opening the file; It has the defer method that accepts a callback function which is suitable for clean-up that will be invoked at the moment of disposing the `cleanup` object.  
```

remypar5 commented on data/blog/explicit-resource-management-in-typescript-5.2.md:
````diff
@@ -0,0 +1,228 @@
+```
+
+Here we’ve created an instance of `DisposableStack` right after opening the file; It has the defer method that accepts a callback function which is suitable for clean-up that will be invoked at the moment of disposing the clean-up object.  
+After the above code is compiled, the resulting JavaScript code will include a `finally` block. This `finally` block will call the callback function that was passed to the defer method. This ensures that the callback function is always executed, even if an error occurs in the try block.
````
```suggestion
After the above code is compiled, the resulting JavaScript code will include a `finally` block. This `finally` block will call the callback function that was passed to the `defer` method. This ensures that the callback function is always executed, even if an error occurs in the try block.
```

remypar5 commented on data/blog/explicit-resource-management-in-typescript-5.2.md:
```diff
@@ -0,0 +1,228 @@
+
+### Async dispose method
+
+Sometimes for your **dispose** logic you might need to use asynchronous operations using `async/await`. In this case, you can simply use the await before the using keyword:
```
```suggestion
Sometimes for your **dispose** logic you might need to use asynchronous operations using `async/await`. In this case, you can simply use the `await` before the `using` keyword:
```

remypar5 reviewed (commented):
I love this article. Today I learned about Explicit Resource Management and its existence.

bagherani committed:
keywords wrapped in code snippets.

bagherani committed:
code and link markdowns splitted to prevent style overwriting

bagherani commented:
> I love this article. Today I learned about Explicit Resource Management and its existence.

Thank you very much, It is very good to hear that. @remypar5 
Nice catches and thanks for your attention.
I fixed the comments. also, to prevent style conflicts, I moved the split the codes and links in the c#, python, and java parts.

remypar5 commented on data/blog/explicit-resource-management-in-typescript-5.2.md:
```diff
@@ -0,0 +1,228 @@
+
+# ECMAScript Explicit Resource Management early implementation in TypeScript 5.2
+
+This article on ECMAScript Explicit Resource Management implementation in TypeScript 5.2 should be learned by developers who are interested in managing resources explicitly in their JavaScript programs. It explains how this feature works and how it can simplify resource management code.
```
```suggestion
This article on "ECMAScript Explicit Resource Management early implementation in TypeScript 5.2" is for developers who are interested in managing resources explicitly in their JavaScript programs. It explains how this feature works and how it can simplify resource management code.
```

bagherani replied:
Thank you for the insightful and accurate look.
fixed

remypar5 commented on data/blog/explicit-resource-management-in-typescript-5.2.md:
```diff
@@ -0,0 +1,228 @@
+
+A "Resource" is something that has a specific lifespan. When that lifespan is over, the program needs to perform a specific action, like closing a file or freeing up memory, so that the program can continue running smoothly. Examples of resources include file handles and network sockets.
+
+At the moment of writing this article, the ECMAScript [proposal](https://github.com/tc39/proposal-explicit-resource-management) for Explicit Resource Management is at [stage 3](https://tc39.es/process-document/) which means it will be launched within a few months and this is the time that the TypeScript community has already implemented it in TypeScript [version 5.2](https://devblogs.microsoft.co ...
```
```suggestion
At the moment of writing, the ECMAScript [proposal](https://github.com/tc39/proposal-explicit-resource-management) for Explicit Resource Management is at [stage 3](https://tc39.es/process-document/) which means it will be launched within a few months. However, the TypeScript community has already implemented it in TypeScript [version 5.2](https://devblogs.microsoft.com/typescript/announcing-typescript-5-2-beta/#using-declarations-and-explicit-resource-management).
```

Break this sentence into 2 smaller ones

bagherani replied:
Done

remypar5 commented on data/blog/explicit-resource-management-in-typescript-5.2.md:
```diff
@@ -0,0 +1,228 @@
+
+At the moment of writing this article, the ECMAScript [proposal](https://github.com/tc39/proposal-explicit-resource-management) for Explicit Resource Management is at [stage 3](https://tc39.es/process-document/) which means it will be launched within a few months and this is the time that the TypeScript community has already implemented it in TypeScript [version 5.2](https://devblogs.microsoft.co ...
+
+This implementation will add a feature to the JavvaScript language that we had similar to in other languages such as C# `using` syntax [[link](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/statements#1314-the-using-statement)], Python `with` syntax [[link](https://docs.python.org/3/reference/compound_stmts.html#the-with-statement)], or `try` [[link](htt ...
```
```suggestion
This implementation will add a feature to the JavaScript language similar to other languages such as C# `using` syntax [[link](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/statements#1314-the-using-statement)], Python `with` syntax [[link](https://docs.python.org/3/reference/compound_stmts.html#the-with-statement)], or `try` [[link](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html)] in Java.
```

bagherani replied:
Done. nice catch.

remypar5 commented on data/blog/explicit-resource-management-in-typescript-5.2.md:
````diff
@@ -0,0 +1,228 @@
+}
+```
+
+As you can see, here the TypeScript compiler has compiled the using syntax into the traditional `try/finally` pattern that I mentioned before and of course, with some simple functions to manage the allocated resources(`__addDisposableResource` and `__disposeResources` will be generated by the compiler in the output file).
````
```suggestion
As you can see here the TypeScript compiler has compiled the using syntax into the traditional `try/finally` pattern that I mentioned before and of course, with some simple functions to manage the allocated resources(`__addDisposableResource` and `__disposeResources` will be generated by the compiler in the output file).
```

bagherani replied:
fixed

bagherani committed:
keywords wrapped with code, phrases improved

sanderdejong88 committed:
Update publish date
